File: leetcode/array/combination_sum_BACKTRACKING.go
package main

import "sort"

func combinationSum(arr []int, target int) [][]int {
	result := make([][]int, 0)
	sort.Ints(arr)

	temp := []int{}
	backTrack(arr, temp, &result, target, 0)
	return result
}

func backTrack(arr []int, temp []int, result *[][]int, remain, start int) {
	if remain < 0 {
		return
	}
	if remain == 0 {
		r := append([]int{}, temp...)
		*result = append(*result, r)
		return
	}

	for i := start; i < len(arr); i++ {
		t := append(temp, arr[i])
		// Since the problem allows for solutions, to have duplicates, for start,
		// we use `i`, thus we start from the same index, in the recursion.
		backTrack(arr, t, result, remain-arr[i], i)
	}
}

File: leetcode/array/combination_sum_BACKTRACKING.py
def combinationSum(arr, target):
    result = []
    temp = []

    backtrack(result, temp, arr, target, 0)
    return result


def backtrack(result, temp, arr, remain, start):
    if remain < 0:
        return
    
    if remain == 0:
        result.append(list(temp))
        return
    
    for i in range(start, len(arr)):
        temp.append(arr[i])
        backtrack(result, temp, arr, remain-arr[i], i) # not i + 1, since we can reuse same element
        temp.pop()


print combinationSum([2, 3, 6, 7], 7)
